packages/block-editor/src/hooks/border-color.js

```diff
@@ -0,0 +1,197 @@
+
+const BORDER_COLOR_SUPPORT_KEY = 'color';
+const EMPTY_ARRAY = [];
+
+/**
+ * Inspector control panel containing the border color related configuration.
+ *
+ * There is deliberate overlap between the colors and borders block supports
+ * relating to border color. It can be argued the border color controls could
+ * be included within either, or both, the colors and borders panels in the
+ * inspector controls. If they share the same block attributes it should not
+ * matter.
+ *
+ * @param  {Object} props Block properties.
+ * @return {WPElement}    Border color edit element.
+ */
+export function BorderColorEdit( props ) {
+	const {
+		attributes: { borderColor, style },
+		setAttributes,
+	} = props;
+	const colors = useEditorFeature( 'color.palette' ) || EMPTY_ARRAY;
+
+	const disableCustomColors = ! useEditorFeature( 'color.custom' );
+	const disableCustomGradients = ! useEditorFeature( 'color.customGradient' );
+
+	if ( useIsBorderColorDisabled( props ) ) {
+		return null;
+	}
+
+	const onChangeColor = ( value ) => {
+		const colorObject = getColorObjectByColorValue( colors, value );
+		const newStyle = {
+			...style,
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
+			},
+		};
+
+		const newNamedColor = colorObject?.slug ? colorObject.slug : undefined;
+
+		setAttributes( {
+			style: cleanEmptyObject( newStyle ),
+			borderColor: newNamedColor,
+		} );
+	};
+
+	return (
+		<ColorGradientControl
+			label={ __( 'Border color' ) }
+			value={ borderColor || style?.border?.color }
+			disableCustomColors={ disableCustomColors }
+			disableCustomGradients={ disableCustomGradients }
+			onColorChange={ onChangeColor }
+		/>
+	);
+}
+
+/**
+ * Determines if there is border color support.
+ *
+ * @param  {string|Object} blockType Block name or Block Type object.
+ * @return {boolean}                 Whether there is support.
+ */
+export function hasBorderColorSupport( blockType ) {
+	const support = getBlockSupport( blockType, BORDER_SUPPORT_KEY );
+	return !! ( true === support || support?.color );
+}
+
+/**
+ * Custom hook that checks if border color settings have been disabled.
+ *
+ * @param  {string} name The name of the block.
+ * @return {boolean}     Whether border color setting is disabled.
+ */
+export function useIsBorderColorDisabled( { name: blockName } = {} ) {
+	const isDisabled = ! useEditorFeature( 'border.customColor' );
+	return ! hasBorderColorSupport( blockName ) || isDisabled;
+}
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `borderColor` if needed.
+ *
+ * @param  {Object} settings Original block settings.
+ * @return {Object}          Updated block settings.
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, BORDER_COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// Allow blocks to specify default value if needed.
+	if ( ! settings.attributes.borderColor ) {
+		Object.assign( settings.attributes, {
```

Comment:
changing the variable passed as an argument is a bad practice. It is better to return a new object with updated value

Reply:
Thanks, I've updated this to return a new settings object including the new `borderColor` attribute definition.